File: apps/web/src/lib/api/subscriber.ts
import { apiClient } from "./client";
import type { ApiResponse } from "./types";

// Subscriber 타입 정의
export interface Subscriber {
  id: string;
  email: string;
  status: "PENDING" | "ACTIVE" | "UNSUBSCRIBED";
  subscribedAt: string;
  confirmedAt: string | null;
  unsubscribedAt: string | null;
  newsletterId: string;
}

// 구독 확인 API 전용 응답
export interface SubscribeConfirmResponse {
  id: string;
  email: string;
  status: "ACTIVE";
  newsletterId: string;
  username: string; // 뉴스레터 소유자 username
  newsletterSlug: string; // 뉴스레터 slug
}

// 구독자 생성 요청
export interface CreateSubscriberRequest {
  email: string;
}

// 구독자 목록 조회
export async function getSubscribers(
  newsletterId: string,
  status?: "PENDING" | "ACTIVE" | "UNSUBSCRIBED"
): Promise<Subscriber[]> {
  const params = new URLSearchParams();
  if (status) {
    params.append("status", status);
  }
  const queryString = params.toString();
  const url = `/newsletters/${newsletterId}/subscribers${queryString ? `?${queryString}` : ""}`;
  
  const response = await apiClient.get<ApiResponse<Subscriber[]>>(url);
  if (!response.data.data) {
    throw new Error(response.data.message || "Failed to get subscribers");
  }
  return response.data.data;
}

// 구독자 추가
export async function createSubscriber(
  newsletterId: string,
  data: CreateSubscriberRequest
): Promise<Subscriber> {
  const response = await apiClient.post<ApiResponse<Subscriber>>(
    `/newsletters/${newsletterId}/subscribers`,
    data
  );
  if (!response.data.data) {
    throw new Error(response.data.message || "Failed to create subscriber");
  }
  return response.data.data;
}

// 구독자 삭제
export async function deleteSubscriber(
  newsletterId: string,
  subscriberId: string
): Promise<void> {
  const response = await apiClient.delete<ApiResponse<void>>(
    `/newsletters/${newsletterId}/subscribers/${subscriberId}`
  );
  if (!response.data.success) {
    throw new Error(response.data.message || "Failed to delete subscriber");
  }
}

// Public API for subscribe (no JWT required)
export interface PublicSubscribeRequest {
  email: string;
}

export async function publicSubscribe(
  newsletterId: string,
  data: PublicSubscribeRequest
): Promise<Subscriber> {
  const response = await apiClient.post<ApiResponse<Subscriber>>(
    `/public/newsletter/${newsletterId}/subscribe`,
    data
  );
  if (!response.data.data) {
    const error = new Error(response.data.message || "Failed to subscribe");
    (error as any).response = response;
    throw error;
  }
  return response.data.data;
}

export async function confirmSubscribe(
  token: string
): Promise<SubscribeConfirmResponse> {
  const response = await apiClient.get<ApiResponse<SubscribeConfirmResponse>>(
    `/public/subscribe/confirm?token=${token}`
  );
  if (!response.data.data) {
    throw new Error(response.data.message || "Failed to confirm subscribe");
  }
  return response.data.data;
}

// 구독 취소 요청
export interface UnsubscribeRequest {
  email: string;
}

export interface UnsubscribeResponse {
  id: string;
  email: string;
  status: "UNSUBSCRIBED";
  newsletterId: string;
}

/**
 * 구독 취소 (Public API - JWT 불필요)
 */
export async function unsubscribe(
  newsletterId: string,
  data: UnsubscribeRequest
): Promise<UnsubscribeResponse> {
  const response = await apiClient.post<ApiResponse<UnsubscribeResponse>>(
    `/public/newsletter/${newsletterId}/unsubscribe`,
    data
  );
  if (!response.data.data) {
    throw new Error(response.data.message || "Failed to unsubscribe");
  }
  return response.data.data;
}

File: apps/api/src/main/kotlin/io/vality/domain/Subscriber.kt
package io.vality.domain

import kotlinx.serialization.Serializable
import kotlinx.serialization.Contextual
import java.time.Instant

/**
 * 구독자 상태 (Subscriber Status)
 * 
 * 뉴스레터를 구독하는 이메일 구독자의 상태를 나타냅니다.
 * 
 * 주의: 이는 뉴스레터 발행자의 Vality 서비스 구독(Subscription)과는 다른 개념입니다.
 * - Subscriber: 뉴스레터를 구독하는 이메일 구독자
 * - Subscription: 뉴스레터 발행자가 Vality 서비스를 사용하기 위한 플랜
 */
enum class SubStatus {
    PENDING,      // 이메일 인증 대기 중 (Double Opt-In)
    ACTIVE,       // 활성 구독자 (인증 완료, 이메일 수신 중)
    UNSUBSCRIBED, // 구독 취소함
}

/**
 * 구독자 (Subscriber)
 * 
 * 이는 뉴스레터를 구독하는 이메일 구독자 정보입니다.
 * 
 * 예시:
 * - 사용자 A가 "기술 뉴스레터"를 이메일로 구독 → Subscriber (newsletterId: "기술 뉴스레터" ID)
 * - 사용자 B가 "디자인 뉴스레터"를 이메일로 구독 → Subscriber (newsletterId: "디자인 뉴스레터" ID)
 * 
 * 주의: 뉴스레터 발행자가 Vality 서비스를 사용하기 위한 플랜 구독(Subscription)과는 완전히 다른 개념입니다.
 * - Subscriber: 뉴스레터를 구독하는 이메일 구독자 (이메일을 받는 사람)
 * - Subscription: 뉴스레터 발행자가 Vality 서비스를 사용하기 위한 플랜 (서비스를 사용하는 사람)
 * 
 * 관계:
 * - User (뉴스레터 발행자) → Subscription (Vality 서비스 플랜 구매)
 * - User (뉴스레터 발행자) → Newsletter (뉴스레터 생성)
 * - Newsletter → Subscriber (이메일 구독자)
 * 
 * @property id 구독자 ID (CUID)
 * @property email 구독자 이메일 주소
 * @property status 구독자 상태 (PENDING, ACTIVE, UNSUBSCRIBED)
 * @property subscribedAt 구독 신청 일시
 * @property confirmedAt 이메일 인증 완료 일시 (Double Opt-In)
 * @property unsubscribedAt 구독 취소 일시
 * @property newsletterId 구독한 뉴스레터 ID
 */
@Serializable
data class Subscriber(
    val id: String,
    val email: String,
    val status: SubStatus = SubStatus.PENDING,
    @Contextual val subscribedAt: Instant,
    @Contextual val confirmedAt: Instant? = null,
    @Contextual val unsubscribedAt: Instant? = null,
    val newsletterId: String,
)
